px.py

```diff
@@ -715,41 +744,111 @@ def do_CONNECT(self):
+                                wdata = cdata
+                                source = "proxy"
+                            else:
+                                out = self.proxy_socket
+                                wdata = sdata
+                                source = "client"
+
+                            data = i.recv(4096)
+                            if data:
+                                dlen = len(data)
+                                cl += dlen
+                                # Prepare data to send it later in the outs section
+                                wdata.append(data)
+                                if out not in outs:
+                                    outs.append(out)
+                                max_idle = time.time() + idle
+                            else:
+                                # No data means connection closed by remote host
+                                dprint("Connection closed by %s" % source)
+                                # Because tunnel is closed on one end there is no need to read from both ends
+                                rlist.clear()
+                                # Do not write anymore to the closed end
+                                if i in wlist:
+                                    wlist.remove(i)
+                                if i in outs:
+                                    outs.remove(i)
+                    if outs:
+                        for o in outs:
+                            if o is self.proxy_socket:
+                                wdata = sdata
+                            else:
+                                wdata = cdata
+                            data = wdata[0]
+                            # socket.send() may sending only a part of the data (as documentation says).
+                            # To ensure sending all data
+                            bsnt = o.send(data)
+                            if bsnt > 0:
+                                if bsnt < len(data):
+                                    # Not all data was sent; store data not sent and ensure select() get's it
+                                    # when the socket can be written again
+                                    wdata[0] = data[bsnt:]
+                                    if o not in wlist:
+                                        wlist.append(o)
+                                else:
+                                    wdata.pop(0)
+                                    if not data and o in wlist:
+                                        wlist.remove(o)
+                                cs += bsnt
+                            else:
+                                dprint("No data sent")
+                        max_idle = time.time() + idle
+                    if max_idle < time.time():
+                        # No data in timeout seconds
+                        dprint("Proxy connection timeout")
+                        break
```

Comment:
Looks right but I'm getting a nagging feeling that this could be more efficient. Do you see any performance implications with this?

Reply:
I use to respect my own nagging feelings because they are often right, so I respect other's too. :-) Do you mean time.time() could by slow? As I use px every day for accessing the internet (browsing, downloading updates and so on) I would say subjectively there is no performance problem. Though did not measure.

Reply:
I agree, let's just test and validate.

Can you split out this select loop into a separate proc? We can then reuse it for GET and POST as well and get closer to fixing issues #47 and #52. Hopefully, it will work the same for a stream (self.rfile) and a socket pair as well.

Reply:
Good idea. It should work I think. But as it needs some major changes (returning something other than the complete http body from some functions) I would rather do it in another change and not within this pull request.

Reply:
Sounds fair.